### PR TITLE
ISSUE-74 : Fix when parsing an XCApi containing different components with the same stateMachine name

### LINKED ISSUE #74: API initialization problem
XComponentApi.CreateFromXCApi throws an exception when different components have the same state machine name

### PR DESCRIPTION
**Pull request by yahyanajar**
closes #74

### CODE REVIEW

File: ReactiveXComponent/Parser/XCApiConfigParser.cs [hunk 1]
```diff
@@ -11,55 +11,61 @@ namespace ReactiveXComponent.Parser
...
-
-        private Dictionary<string, int> _componentCodeByComponent;
-        private Dictionary<string, int> _stateMachineCodeByStateMachine;
+        private XCApiDescription _xcApiDescription;
+        private XNamespace _xc;
         private Dictionary<string, int> _eventCodeByEvent;
         private Dictionary<TopicIdentifier, string> _publisherTopicByIdentifier;
         private Dictionary<TopicIdentifier, string> _subscriberTopicByIdentifier;
         private Dictionary<long, string> _snapshotTopicByComponent;
+        private Dictionary<string, List<StateMachineInfo>> _stateMachineInfoByComponentRepository;
```
Comment (Julien-Molina) [inline ReactiveXComponent/Parser/XCApiConfigParser.cs]: Choose something "Repo" "Repository" or "Nothing" and use it everywhere. I would suggest to remove Repository & Repo.

File: ReactiveXComponent/Parser/XCApiConfigParser.cs [hunk 2]
```diff
@@ -11,55 +11,61 @@ namespace ReactiveXComponent.Parser
...
 
         public void Parse(Stream xcApiStream)
         {
             var reader = XmlReader.Create(xcApiStream);
             _xcApiDescription = new XCApiDescription(reader);
             _xc = "http://xcomponent.com/DeploymentConfig.xsd";
-
-            _componentCodeByComponent = CreateComponentCodeByNameRepository(_xcApiDescription.GetComponentsNode());
-            _stateMachineCodeByStateMachine = CreateStateMachineCodeByNameRepository(_xcApiDescription.GetStateMachinesNode());
+            ParseComponentNodes(_xcApiDescription.GetComponentsNode(), out _componentCodeByComponentNameRepo, out _stateMachineInfoByComponentRepository);
```
Comment (Julien-Molina) [inline ReactiveXComponent/Parser/XCApiConfigParser.cs]: Remove out _componentCodeByComponentNameRepo, out _stateMachineInfoByComponentRepository and use them directly in the private method

File: ReactiveXComponent/Parser/XCApiConfigParser.cs [hunk 3]
```diff
@@ -114,43 +120,40 @@ private void AddTopicToRepository(Dictionary<TopicIdentifier, string> repository
 
             if (!repository.ContainsKey(topicIdentifier))
             {
-                repository.Add(topicIdentifier, node?.Descendants(_xc + XCApiTags.Topic).FirstOrDefault().Value);
+                repository.Add(topicIdentifier, node?.Descendants(_xc + XCApiTags.Topic).FirstOrDefault()?.Value);
             }
         }
 
         private Dictionary<long, string> CreateSnapshotTopicByComponentRepository(IEnumerable<XElement> snapshotNodes)
```
Comment (Julien-Molina) [inline ReactiveXComponent/Parser/XCApiConfigParser.cs]: Same problem with namings here : "Repo" and "Repository"

File: ReactiveXComponent/Parser/XCApiConfigParser.cs [hunk 4]
```diff
@@ -114,43 +120,40 @@ private void AddTopicToRepository(Dictionary<TopicIdentifier, string> repository
...
 
-        private void AddSnapshotTopicToRepository(Dictionary<long, string> repository, XElement node)
+        private void AddSnapshotTopicToRepository(IDictionary<long, string> repository, XElement node)
         {
             var componentCode = Convert.ToInt64(node?.Attribute(XCApiTags.ComponentCode)?.Value);
 
             if (!repository.ContainsKey(componentCode))
             {
-                repository.Add(componentCode, node?.Descendants(_xc + XCApiTags.Topic).FirstOrDefault().Value);
+                repository.Add(componentCode, node?.Descendants(_xc + XCApiTags.Topic).FirstOrDefault()?.Value);
```
Comment (Julien-Molina) [inline ReactiveXComponent/Parser/XCApiConfigParser.cs]: Do we really want to add null values in the dictionnary ?

File: ReactiveXComponent/Parser/StateMachineInfo.cs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+﻿namespace ReactiveXComponent.Parser
+{
+    public class StateMachineInfo
+    {
```
Comment (Julien-Molina) [inline ReactiveXComponent/Parser/StateMachineInfo.cs]: Create a constructor with 2 parameters and remove the public setters

File: ReactiveXComponent/Parser/XCApiConfigParser.cs (hunk 1 above)
Comment (yahyanajar) [reply ReactiveXComponent/Parser/XCApiConfigParser.cs]: done

File: ReactiveXComponent/Parser/StateMachineInfo.cs (hunk 5 above)
Comment (yahyanajar) [reply ReactiveXComponent/Parser/StateMachineInfo.cs]: done

File: ReactiveXComponent/Parser/XCApiConfigParser.cs (hunk 2 above)
Comment (yahyanajar) [reply ReactiveXComponent/Parser/XCApiConfigParser.cs]: done

File: ReactiveXComponent/Parser/XCApiConfigParser.cs (hunk 3 above)
Comment (yahyanajar) [reply ReactiveXComponent/Parser/XCApiConfigParser.cs]: done

**Pull request merged**